src/transformers/trainer.py

```diff
@@ -651,6 +714,51 @@ def train(self, model_path: Optional[str] = None):
         logger.info("\n\nTraining completed. Do not forget to share your model on huggingface.co/models =)\n\n")
         return TrainOutput(self.global_step, tr_loss / self.global_step)
 
+    def hyperparameter_search(
+        self, n_trials: int = 100, timeout: int = 1800, n_jobs: int = 1, direction: str = "minimize", **kwargs
+    ) -> BestRun:
+        """
+        Launch an hyperparameter search using ``optuna``. The optimized quantity is determined by the
+        :meth:`~transformers.Trainer.compute_obejctive` method, which is the evaluation loss when no metric is
+        provided, the sum of all metrics otherwise (you can change that behavior by subclassing and overriding this
+        method).
+
+        Args:
+            n_trial (:obj:`int`, `optional`, defaults to 100):
+                The number of trial runs to test.
+            timeout (:obj:`int`, `optional`, defaults to 1800):
+                The number of seconds before abandoning a trial.
+            n_jobs (:obj:`int`, `optional`, defaults to 1):
+                The number of jobs to launch in parallel.
+            direction(:obj:`str`, `optional`, defaults to :obj:`"minimize"`):
+                Whether to optimize greater or lower objects. Can be :obj:`"minimize"` or :obj:`"maximize"`, you should
+                pick :obj:`"minimize"` when optimizing the validation loss, :obj:`"maximize"` when optimizing one or
+                several metrics.
+            kwargs:
+                Additional keyword arguments passed along to :obj:`optuna.create_study`.
+
+        Returns:
+            :class:`transformers.trainer_utils.BestRun`: All the informations about the best run.
+        """
+        assert is_optuna_available(), "optuna is required to do hyperparameter search, use `pip install optuna`."
+
+        def _objective(self, trial):
+            # To make sure optimizer and lr_scheduler are reset with the new choices of HPs
+            self.optimizer = None
+            self.lr_scheduler = None
```

Comment:
This means the `hyperparameter_search` is not compatible with passing a custom (optimizer, scheduler) to the Trainer, right?

Reply:
Not for now. I don't think you can reset a scheduler in PyTorch so custom `(optimizer, scheduler)` means the user needs to subclass and override the `create_optimizer_and_scheduler` here. Will add to the documentation.

Reply:
Yes, fair enough. I like subclassing as an extension point.